custom_metrics/sass.js

```diff
@@ -0,0 +1,309 @@
+			ret.ifs.push(obj);
+		}
+		else {
+			// New conditional
+			ret.ifs.push({test});
+		}
+	});
+
+	// @each, @for, @while Loops
+	ret.eaches = {};
+	scss.replace(/@each\s+(.+ in .+)(?=\s*\{)/g, ($0, args) => {
+		ret.eaches[args] = (ret.eaches[args] || 0) + 1;
+	});
+
+	ret.fors = {};
+	// console.log("@for", scss);
+	scss.replace(/@for\s+(.+)(?=\s*\{)/g, ($0, args) => {
+		ret.fors[args] = (ret.fors[args] || 0) + 1;
+	});
+
+	ret.whiles = {};
+	scss.replace(/@while\s+(.+)(?=\s*\{)/g, ($0, args) => {
+		ret.whiles[args] = (ret.whiles[args] || 0) + 1;
+	});
+
+	// @extend
+	ret.extends = {};
+	scss.replace(/@extend\s*(.+)(?=\s*(?:[;}]|$))/g, ($0, selector) => {
+		ret.extends[selector] = (ret.extends[selector] || 0) + 1;
+	});
+
+	// @error
+	ret.errors = scss.match(/@error (.+)/g)?.map(e => e.slice(7));
+
+	// CSS variables that are set with Sass variables
+	ret.variablesCombined = {};
+	scss.replace(/(?<=^|\s)--([\w-]+):\s*(.*#\{.*\$.+\}.*)\s*$/gm, ($0, name, value) => {
+		ret.variablesCombined["--" + name] = value;
+	});
+
+	// Heuristic for nesting &
+	ret.nested = {
+		"descendant": scss.match(/&\s+(?=[\w[.:#])/g)?.length,
+		">": scss.match(/&\s*(?=>)/g)?.length,
+		"+": scss.match(/&\s*(?=\+)/g)?.length,
+		"~": scss.match(/&\s*(?=~)/g)?.length,
+		"lone": scss.match(/&\s*(?=[,{]|$})/g)?.length,
+		"pseudo-element": scss.match(/&::/g)?.length,
+		"pseudo-class": scss.match(/&:(?!:)/g)?.length,
+		"id": scss.match(/&#\w+/g)?.length,
+		"class": scss.match(/&\.\w+/g)?.length,
+		"attr": scss.match(/&\[\w+/g)?.length,
+	};
+}
+
+let results = (async () => {
+
+const results = {};
+const $$ = s => [...document.querySelectorAll(s)];
+const sourcemapRegex = /\/\*[#@] sourceMappingURL=(.+?) \*\//;
+
+let stylesheets;
+
+if (typeof $WPT_BODIES !== "undefined") {
+	// in WPT, we already have the bodies
+	stylesheets = $WPT_BODIES.filter(request => request.type == "Stylesheet")
+		.map(file => {
+			return {url: file.url, body: file.response_body};
+		});
+}
+else {
+	// Outside WPT, use fetch()
+	stylesheets = await Promise.all($$("link[rel=stylesheet]").map(async link => {
+		try {
+			let response = await fetch(link.href);
+			let body = await response.text();
+
+			return {url: link.href, body};
+		}
+		catch (e) {}
+	}));
+}
+
+results.stylesheets = {
+	remote: stylesheets.length
+};
+
+stylesheets.push(...$$("style").map(s => {
+	return {
+		url: location,
+		body: s.textContent
+	}
+}));
+
+results.stylesheets.inline = stylesheets.length - results.stylesheets.remote;
+
+let sourcemapURLs = stylesheets.map(o => {
+	if (o) {
+		let url = o.body.match(sourcemapRegex)?.[1];
+
+		if (url) {
+			// Source map URL is relative to stylesheet URL
+			return new URL(url, o.url);
+		}
+	}
+}).filter(url => !!url);
+
+results.sourcemaps = {
+	count: sourcemapURLs.length,
+	ext: {}
+};
+
+if (sourcemapURLs.length === 0) {
+	return results;
+}
+// console.log("sourcemap URLs", sourcemapURLs);
+
+// Assumption: Either all sources are SCSS or none.
```

Comment:
What does this imply? Will it fail if one source is not SCSS?

Reply:
No. If you look at the sample output, many of the sources are a mix of CSS and SCSS and it does fine.
Basically what it means is that we read & join the files if at least one is SCSS.